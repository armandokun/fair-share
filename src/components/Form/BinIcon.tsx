import { DeleteOutlined } from "@mui/icons-material";
import { IconButton, InputAdornment } from "@mui/material";

type Props = {
  position: "start" | "end";
  onClick?: React.MouseEventHandler<HTMLButtonElement>;
};

const BinIcon = ({ position, onClick }: Props) => (
  <InputAdornment position={position}>
    <IconButton onClick={onClick}>
      <DeleteOutlined />
    </IconButton>
  </InputAdornment>
);

export default BinIcon;
