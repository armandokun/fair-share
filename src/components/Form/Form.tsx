import { ChangeEvent, useEffect, useState } from "react";
import {
  Box,
  Button,
  Card,
  Container,
  Divider,
  IconButton,
  InputAdornment,
  Stack,
  TextField,
  Typography,
} from "@mui/material";
import { Add, DeleteOutlined } from "@mui/icons-material";
import { useTranslation } from "react-i18next";

import { REQUIRED_SALARIES } from "../../constants";
import {
  getIndividualPayment,
  getSalaryPercent,
  decimalCount,
} from "../../helpers/form";

const Form = () => {
  const [salaries, setSalaries] = useState<number[]>([0, 0]);
  const [salaryPercent, setSalaryPercent] = useState(0);
  const [rentToPay, setRentToPay] = useState(0);

  const { t } = useTranslation("translation", { keyPrefix: "form" });

  useEffect(() => {
    setSalaryPercent(getSalaryPercent(salaries, rentToPay));
  }, [salaries, rentToPay]);

  const handleChange =
    (index: number) => (event: ChangeEvent<HTMLInputElement>) => {
      const newSalaries = [...salaries];

      newSalaries[index] = Number(event.target.value);
      setSalaries(newSalaries);
    };

  const handleRemove = (index: number) => () =>
    setSalaries((prevState) => {
      prevState.splice(index, 1);
      return [...prevState];
    });

  const renderBinIcon = (index: number) => (
    <InputAdornment position="end">
      <IconButton onClick={handleRemove(index)}>
        <DeleteOutlined />
      </IconButton>
    </InputAdornment>
  );

  const renderSalariesInputs = () => {
    return salaries.map((_, index) => {
      return (
        <TextField
          key={`salary-${index}`}
          label={t("salary.title", { number: index + 1 })}
          type="number"
          InputProps={{
            endAdornment: index >= REQUIRED_SALARIES && renderBinIcon(index),
          }}
          required={index < REQUIRED_SALARIES}
          onChange={handleChange(index)}
        />
      );
    });
  };

  const renderEvaluationNote = () => {
    if (salaryPercent > 99) return t("evaluation_note.title.error");
    if (salaryPercent === 50) return t("evaluation_note.title.half");
    if (salaryPercent > 50) return t("evaluation_note.title.too_much");
    if (salaryPercent < 10) return t("evaluation_note.title.cheap");

    return null;
  };

  const renderSummary = () => {
    const formattedSalaryPercent = salaryPercent.toPrecision(2);
    const summarySalaryPercentText =
      decimalCount(salaryPercent) === 0
        ? t("summary.description_exact", {
            salaryPercent: formattedSalaryPercent,
          })
        : t("summary.description_around", {
            salaryPercent: formattedSalaryPercent,
          });

    return (
      <Container sx={{ mb: 2 }}>
        <Box sx={{ "& .MuiTextField-root": { m: 2 } }}>
          <Typography variant="h6">{t("summary.title")}</Typography>
          <Typography variant="body1" sx={{ mb: 2 }}>
            {summarySalaryPercentText}
          </Typography>
          {salaries.map((salary, index) => {
            return (
              <Typography variant="body1" key={`summary-${index}`}>
                {t("summary.salary.title", {
                  number: index + 1,
                  payment: getIndividualPayment(salary, salaryPercent),
                })}
              </Typography>
            );
          })}
          <Typography variant="body2" sx={{ mt: 2 }}>
            {renderEvaluationNote()}
          </Typography>
        </Box>
      </Container>
    );
  };

  return (
    <Card variant="outlined" sx={{ m: 3 }}>
      <Box
        component="form"
        sx={{
          "& .MuiTextField-root": { m: 2, width: "25ch" },
        }}
      >
        <Button
          variant="outlined"
          startIcon={<Add />}
          onClick={() => setSalaries((prevState) => [...prevState, 0])}
          sx={{ mt: 2, ml: 2 }}
        >
          {t("actions.add")}
        </Button>
        <Stack spacing={1}>{renderSalariesInputs()}</Stack>
        <Divider variant="middle" />
        <TextField
          required
          id="outlined-required"
          label={t("rent.title")}
          type="number"
          onChange={(e) => setRentToPay(Number(e.target.value))}
        />
        {!!salaryPercent && renderSummary()}
      </Box>
    </Card>
  );
};

export default Form;
