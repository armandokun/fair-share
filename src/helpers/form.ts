export const getSalaryPercent = (
  salaries: Array<number>,
  rentToPay: number
) => {
  const totalSalary = salaries.reduce(
    (accumulator, currentNumber) => accumulator + currentNumber,
    0
  );

  return (rentToPay / totalSalary) * 100;
};

export const getIndividualPayment = (
  salary: number,
  salaryPercent: number
) => {
  const individualPayment = (salaryPercent / 100) * salary

  return individualPayment.toFixed(2)
};

export const decimalCount = (number: number) => {
  const numberAsString = number.toString();

  if (numberAsString.includes('.')) {
    return numberAsString.split('.')[1].length;
  }

  return 0;
}
