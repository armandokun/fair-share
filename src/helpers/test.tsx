import { I18nextProvider } from "react-i18next";

import i18n from "../translations";

export const withIntl = (children: JSX.Element) => {
  return <I18nextProvider i18n={i18n}>{children}</I18nextProvider>;
};
