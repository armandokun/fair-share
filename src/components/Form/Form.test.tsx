import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import Form from ".";
import { withIntl } from "../../helpers/test";

describe("<Form />", () => {
  const renderHelper = (element: JSX.Element) => render(withIntl(element));

  it("renders correctly", () => {
    const { container } = renderHelper(<Form />);

    expect(container).toMatchSnapshot();
  });

  it("adds salary input when clicked on add button", () => {
    renderHelper(<Form />);

    userEvent.click(screen.getByText("Add another person"));

    expect(screen.getByLabelText("Salary 3")).toBeInTheDocument();
  });

  it("removes salary input when clicked on remove button", () => {
    renderHelper(<Form />);

    userEvent.click(screen.getByText("Add another person"));
    userEvent.click(screen.getByTestId("remove-button"));

    expect(screen.queryByLabelText("Salary 3")).not.toBeInTheDocument();
  });

  it("does not render summary when only rent input is filled", () => {
    renderHelper(<Form />);

    userEvent.type(screen.getByLabelText("Rent to pay *"), "100");

    expect(screen.queryByText("Summary")).not.toBeInTheDocument();
  });

  it("renders summary when rent and salary inputs are filled", () => {
    renderHelper(<Form />);

    userEvent.type(screen.getByLabelText("Rent to pay *"), "100");
    userEvent.type(screen.getByLabelText("Salary 1 *"), "1000");
    userEvent.type(screen.getByLabelText("Salary 2 *"), "1000");

    expect(screen.getByText("Summary")).toBeInTheDocument();
  });

  describe("when summary is rendered", () => {
    const fillInputs = (
      rentToPay: string,
      salary1: string,
      salary2: string
    ) => {
      userEvent.type(screen.getByLabelText("Rent to pay *"), rentToPay);
      userEvent.type(screen.getByLabelText("Salary 1 *"), salary1);
      userEvent.type(screen.getByLabelText("Salary 2 *"), salary2);
    };

    it("renders the salary percent", () => {
      renderHelper(<Form />);

      fillInputs("100", "1000", "1000");

      expect(
        screen.getByText("You're all be paying 5.0% of your salary on rent")
      ).toBeInTheDocument();
    });

    it("renders individual payment to make", () => {
      renderHelper(<Form />);

      fillInputs("100", "1000", "1000");

      expect(screen.getByText("Salary 1 to pay: 50.00")).toBeInTheDocument();
      expect(screen.getByText("Salary 2 to pay: 50.00")).toBeInTheDocument();
    });

    describe("when evaluation note is rendered", () => {
      it("renders note for rent price too high", () => {
        renderHelper(<Form />);

        fillInputs("600", "400", "500");

        expect(
          screen.getByText("Do you really want to rent this place? ðŸ¤”")
        ).toBeInTheDocument();
      });

      it("renders note for rent price being low", () => {
        renderHelper(<Form />);

        fillInputs("100", "1000", "1000");

        expect(
          screen.getByText(
            "Do you want to tell us your secrets about the place you rent?",
            { exact: false }
          )
        ).toBeInTheDocument();
      });

      it("renders note for rent price being 50% of your salaries", () => {
        renderHelper(<Form />);

        fillInputs("1000", "1000", "1000");

        expect(
          screen.getByText("Jackpot! ðŸŽ° You waste 50% of your salary on rent.")
        ).toBeInTheDocument();
      });
    });
  });
});
