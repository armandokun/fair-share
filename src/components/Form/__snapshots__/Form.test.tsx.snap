// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form /> renders correctly 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root css-1rmt08s-MuiPaper-root-MuiCard-root"
  >
    <form
      class="MuiBox-root css-17i7n9v"
    >
      <button
        class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root  css-vup0mh-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="AddIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
            />
          </svg>
        </span>
        Add another person
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <div
        class="css-1okj3ks-MuiStack-root"
      >
        <div
          class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="mui-1"
            id="mui-1-label"
          >
            Salary 1
            <span
              aria-hidden="true"
              class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
            >
               
              *
            </span>
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id="mui-1"
              required=""
              type="number"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-1ftyaf0"
              >
                <span>
                  Salary 1
                   
                  *
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="mui-2"
            id="mui-2-label"
          >
            Salary 2
            <span
              aria-hidden="true"
              class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
            >
               
              *
            </span>
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id="mui-2"
              required=""
              type="number"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-1ftyaf0"
              >
                <span>
                  Salary 2
                   
                  *
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <hr
        class="MuiDivider-root MuiDivider-middle css-ft4jij-MuiDivider-root"
      />
      <div
        class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for="outlined-required"
          id="outlined-required-label"
        >
          Rent to pay
          <span
            aria-hidden="true"
            class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id="outlined-required"
            required=""
            type="number"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-1ftyaf0"
            >
              <span>
                Rent to pay
                 
                *
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </form>
  </div>
</div>
`;
